core/src/main/java/org/apache/iceberg/stats/StatsReader.java

```diff
@@ -0,0 +1,183 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.stats;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.io.ByteStreams;
+
+public class StatsReader implements Closeable {
+
+  private final long fileLength;
+  private final SeekableInputStream input;
+  private Optional<Integer> knownFooterSize;
+  private Optional<FileMetadata> knownFileMetadata = Optional.empty();
+
+  public StatsReader(InputFile inputFile, Optional<Integer> footerSize) {
+    this.fileLength = inputFile.getLength();
+    this.input = Preconditions.checkNotNull(inputFile, "inputFile is null").newStream();
+    this.knownFooterSize = Preconditions.checkNotNull(footerSize, "footerSize is null");
+    footerSize.ifPresent(size ->
+        Preconditions.checkArgument(0 < size && size <= fileLength - StatsFormat.getMagic().length,
+            "Invalid footer size: %s", size));
+  }
+
+  public FileMetadata getFileMetadata() throws IOException {
+    if (!knownFileMetadata.isPresent()) {
+      int footerSize = getFooterSize();
+      byte[] footer = readInput(fileLength - footerSize, footerSize);
+
+      checkMagic(footer, 0);
+      checkMagic(footer, footerSize - 4);
+
+      int fileFormatVersion = StatsFormat.readIntegerLittleEndian(footer, footerSize - 8);
+      Preconditions.checkState(
+          fileFormatVersion == StatsFormat.CURRENT_FORMAT_VERSION,
+          "Unsupported format version %s, %s is the latest supported",
+          fileFormatVersion,
+          StatsFormat.CURRENT_FORMAT_VERSION);
+
+      int flags = StatsFormat.readIntegerLittleEndian(footer, footerSize - 12);
+      Preconditions.checkState((flags & ~StatsFormat.SUPPORTED_FLAGS) == 0, "Unsupported flags: %s", flags);
+      boolean compressed = isFlagSet(flags, StatsFormat.FLAG_COMPRESSED);
+
+      int reserved = StatsFormat.readIntegerLittleEndian(footer, footerSize - 16);
+      Preconditions.checkState(reserved == 0, "Unexpected reserved bytes value: %s", reserved);
```

Comment:
This breaks forward compatibility if we ever decide to use the reserved bytes, which defeats the purpose. If we have to increment the format version to use them, we should just remove them.

Reply:
> This breaks forward compatibility

Yes it does, intentionally. A reader cannot read a file version it doesn't support.
This is same as with unrecognized flags. If there is a flag set a reader doesn't know meaning of, it cannot proceed.

> If we have to increment the format version to use them, we should just remove them.

It exists as reserved so that it's easier to add use of it.
Footer has a fixed-layout portion which is hard to evolve, i.e. hard to maintain support for multiple different layouts in same code base. This is why this field is added from the start.